### PR TITLE
Py: Add possibility to check noise keys

### PR DESCRIPTION
**Pull request by TheCharlatan**
Add two arguments to the bitbox02 initialization that allow to pass in a private key and an existing set of static device public keys.

### CODE REVIEW

File: py/bitbox02/communication/bitbox_api_protocol.py [hunk 1]
```diff
@@ -240,6 +246,8 @@ def __init__(
             else:
                 self.close()
                 raise Exception("unexpected response")
+        if remote_static_key not in remote_keys:
+            remote_keys.append(remote_static_key)
```
Comment (benma) [inline py/bitbox02/communication/bitbox_api_protocol.py]: this does not persist it anywhere. maybe it would be easier to copy how it is done in bitbox02-api-go, e.g. pass an instance with some functions for getting/setting keys?

Comment (NickeZ) [reply py/bitbox02/communication/bitbox_api_protocol.py]: I think the application (not the communication library) should store any state. So I think this PR is correct.

**Comment by TheCharlatan** [comment]
I re-wrote this change. It now splits up the `__init__` function and adds a class that holds all the required callbacks. Everything is still optional, because I don't want to re-write all the clients using this.
The callbacks should mimic what is going in `bitbox02-api-go`, please double check if the types in the callbacks all make sense.

**Comment by NickeZ** [comment]
I don't like that there are callbacks. That will lead to spaghetti code in the application code. I liked the first version of this PR more. The go api is not a good example of a clean API. Callbacks are a necessary evil sometimes, but they add a lot of mental overhead, so should be avoided.

**Comment by benma** [comment]
@NickeZ 

I agree that callbacks can be not so nice.

> I think the application (not the communication library) should store any state.

The application is still doing the actual storing.

The previous way with the list append has some api disadvantages (which may be overkill to discuss):

- List[] as a type that is changed by reference is not nearly as tight as it could be. The client does not see in the type signature if only one element is appended, if elements can be deleted, and so on. types should ideally allow as little unwanted behavior as possible.
- the private key generation in the previous version was done by the lib client, but it definitely should be done by the library
- Unusual (?) in Python to have an `inout` argument. I was confused there for a second.

I suggested to use a more common Python idioms: inheritance or class mixins (default implementation as class methods, inherit to change them), but then also suggested to just follow the current style, as there are already callbacks. Can convert from one style to another later, but imho consistent style is good to have.

tl;dr: I'd still go with callbacks (agreeing that they are not very nice), followed by going with `self.contians_device_static_key()` etc. and inherit the class at call.

File: py/bitbox02/communication/bitbox_api_protocol.py [hunk 2]
```diff
@@ -197,26 +215,43 @@ def __init__(
...
         if self._query(OP_I_CAN_HAS_HANDSHAEK) != RESPONSE_SUCCESS:
             self.close()
             raise Exception("Couldn't kick off handshake")
 
         # init noise channel
         noise = NoiseConnection.from_name(b"Noise_XX_25519_ChaChaPoly_SHA256")
         noise.set_as_initiator()
-        dummy_private_key = os.urandom(32)
-        noise.set_keypair_from_private_bytes(Keypair.STATIC, dummy_private_key)
+        if noise_config is not None:
```
Comment (benma) [inline py/bitbox02/communication/bitbox_api_protocol.py]: if it is none there is no private_key?

**Comment by NickeZ** [comment]
If you want to keep the config class then I suggest that you write it as a python class. Basically the four functions should be four functions and the application can extend it to provide functionality.

> the private key generation in the previous version was done by the lib client, but it definitely should be done by the library

Why? In that case I think the library should provide a `generate_privkey(): bytes` function so that the application can create a privkey before it calls the library.

> `contains_device_static_pubkey_callback: Callable[[bytes], bool]`

This function is reasonable, but I think it would be more pythonic to have a `Generator` and then call `if pubkey in keys`.

**Comment by NickeZ** [comment]
But I also agree that there shouldn't be any `inout` variables. So I guess we might need a function for `register_pubkey`.

**Comment by NickeZ** [comment]
Other things I noticed is that when these functions are made member functions they don't have to be named `_callback` and to indicate "no value" `None` should be used and not `b""`. `get_app_static_privkey_callback() == b""` Basically the return value should be `Optional[bytes]`.

**Comment by benma** [comment]
:+1: on class methods over callbacks in the config class.

> >    contains_device_static_pubkey_callback: Callable[[bytes], bool]

> This function is reasonable, but I think it would be more pythonic to have a Generator and then call if pubkey in keys.

I think you might have misread the signature. What's passed is ine pubkey `bytes`, result is `bool`. `[bytes]` looks misleading but I think it's just a function with one argument, `bytes`. The whole function already returns if if the pubkey is present.

**Comment by NickeZ** [comment]
> +1 on class methods over callbacks in the config class.
> 
> > > contains_device_static_pubkey_callback: Callable[[bytes], bool]
> 
> > This function is reasonable, but I think it would be more pythonic to have a Generator and then call if pubkey in keys.
> 
> I think you might have misread the signature. What's passed is ine pubkey `bytes`, result is `bool`. `[bytes]` looks misleading but I think it's just a function with one argument, `bytes`. The whole function already returns if if the pubkey is present.

I don't think I misunderstood, and it is fine to have a "contains" method. But what I meant was that the config object has a method `pubkeys(): Generator[bytes]` and that the library can call `if key in config.pubkeys()`. But obviously it is also OK to have every application instead of the API layer implement that logic.

File: py/bitbox02/communication/bitbox_api_protocol.py (hunk 2 above)
Comment (TheCharlatan) [reply py/bitbox02/communication/bitbox_api_protocol.py]: Whoops, the private_key is now always generated, but not used if the config provides one.
